File: internal/api/nodes.go
package api

import (
	"context"
	"database/sql"
	"fmt"
	"strconv"
	"strings"

	pb "github.com/nagiflow/nagipath/internal/api/pb/nagipath/api/v1"
	"github.com/nagiflow/nagipath/internal/store"
	"github.com/nagiflow/nagipath/internal/trace"
)

// Every nodes.go handler moved to nodeservice.go as NodeService's RPCs
// (docs/adr/0018, proto/nagipath/api/v1/nodes.proto). Everything below
// stays here: shared with nodeservice.go.

// listCap bounds how many rows ListNodes returns in one call, same as
// internal/web/nodes.go's listCap.
const listCap = 500

func filterNodeRows(nodes []store.NodeListRow, q string) []store.NodeListRow {
	q = strings.ToLower(q)
	out := make([]store.NodeListRow, 0, len(nodes))
	for _, n := range nodes {
		if strings.Contains(strings.ToLower(n.DisplayName), q) ||
			strings.Contains(strings.ToLower(n.Address), q) {
			out = append(out, n)
		}
	}
	return out
}

// ---------------------------------------------------------------- node detail

// nodeBuild is GetNode's working accumulator — plain store/trace types,
// converted to proto/nagipath/api/v1/nodes.proto's NodeDetailResponse
// (docs/adr/0018) once, at the end, by toProto(). Keeps the tab-loading
// logic below identical to internal/web/nodes.go's.
type nodeBuild struct {
	Node           store.Node
	Credentials    []store.Credential
	Instances      []store.Instance
	Selected       *store.Instance
	HostKeys       []store.HostKey
	PendingKeys    []store.PendingHostKey
	Stats          store.NodeStats
	Running        bool
	Threshold      int
	Tabs           []nodeTab
	Tab            string
	CredentialName string
	BastionName    string

	Inst         *trace.Inst
	Upstreams    []store.UpstreamPool
	SelectedPool *store.UpstreamPool
	PoolMembers  []store.UpstreamMember
	Certificates []nodeCertBindingBuild
	Files        []store.FileRef
	Snapshot     store.Snapshot
	SelectedFile *store.FileRef
	FileBody     string
	ByteStart    int
	LineStart    int

	DriftRuns     []store.DriftRun
	DriftFindings []store.DriftFinding
	DriftByObject map[string]int
	HasDrift      bool
	NoDriftReason string

	SelectedSiteName string
	SelectedRoute    *trace.Route
	RouteEffect      *routeEffectBuild
}

type nodeTab struct {
	Label string
	Href  string
	Count int
	On    bool
}

type nodeCertBindingBuild struct {
	SubjectCN   string
	SiteName    string
	Port        int
	NotAfter    string
	Fingerprint string
	KeyType     string
	ChainLength int
}

type routeEffectBuild struct {
	UpstreamName  string
	MemberCount   int
	BalanceMethod string
	Members       []*trace.Member
}

func toPBNode(n store.Node) *pb.Node {
	return &pb.Node{
		Id: n.ID, Address: n.Address, SshPort: int32(n.SSHPort), DisplayName: n.DisplayName,
		SshUsername: n.SSHUsername.String, CredentialId: n.CredentialID.Int64, OsFamily: n.OSFamily,
		SudoAvailable: n.SudoAvailable, Enabled: n.Enabled, Source: n.Source, Notes: n.Notes,
		ConsecutiveFailures: int32(n.ConsecutiveFailures), FirstSeenAt: n.FirstSeenAt,
		LastCollection: n.LastCollection.String, LastStatus: n.LastStatus.String,
		BastionNodeId: n.BastionNodeID.Int64,
	}
}

func toPBInstance(in store.Instance) *pb.Instance {
	return &pb.Instance{
		Id: in.ID, NodeId: in.NodeID, ClusterId: in.ClusterID.Int64, Vendor: in.Vendor,
		DisplayName: in.DisplayName, Version: in.Version, MainConfigPath: in.MainConfigPath,
		NodeDisplayName: in.NodeDisplayName, SiteCount: int32(in.SiteCount), RouteCount: int32(in.RouteCount),
		CertCount: int32(in.CertCount), LastCaptured: in.LastCaptured.String, ClusterName: in.ClusterName,
		Degraded: in.Degraded, ParseState: in.ParseState, State: in.State(),
		RestartCommand: effectiveRestartCommand(in),
	}
}

func toPBUpstreamMemberRefs(members []*trace.Member) []*pb.UpstreamMemberRef {
	out := make([]*pb.UpstreamMemberRef, 0, len(members))
	for _, m := range members {
		out = append(out, &pb.UpstreamMemberRef{Id: m.ID, Host: m.Host, Port: int32(m.Port), Scheme: m.Scheme, Flags: m.Flags})
	}
	return out
}

func toPBRoute(rt *trace.Route) *pb.Route {
	if rt == nil {
		return nil
	}
	out := &pb.Route{
		Id: rt.ID, MatchType: rt.MatchType, Pattern: rt.Pattern, Ordinal: int32(rt.Ordinal),
		IsTerminal: rt.IsTerminal, UpstreamId: rt.UpstreamID.Int64, TargetRaw: rt.TargetRaw,
	}
	for _, c := range rt.Children {
		out.Children = append(out.Children, toPBRoute(c))
	}
	return out
}

func toPBInst(inst *trace.Inst) *pb.Inst {
	if inst == nil {
		return nil
	}
	out := &pb.Inst{
		Id: inst.ID, NodeName: inst.NodeName, Vendor: inst.Vendor, DisplayName: inst.DisplayName,
		ClusterName: inst.ClusterName, Degraded: inst.Degraded,
	}
	for _, site := range inst.Sites {
		s := &pb.Site{Id: site.ID, PrimaryName: site.PrimaryName, Kind: site.Kind}
		for _, n := range site.Names {
			s.Names = append(s.Names, n.Name)
		}
		for _, rt := range site.Routes {
			s.Routes = append(s.Routes, toPBRoute(rt))
		}
		out.Sites = append(out.Sites, s)
	}
	for _, up := range inst.Upstreams {
		out.Upstreams = append(out.Upstreams, &pb.Upstream{
			Id: up.ID, Name: up.Name, Kind: up.Kind, BalanceMethod: up.BalanceMethod,
			Members: toPBUpstreamMemberRefs(up.Members),
		})
	}
	return out
}

func (nb *nodeBuild) toProto() *pb.NodeDetailResponse {
	resp := &pb.NodeDetailResponse{
		Node: toPBNode(nb.Node), Running: nb.Running, Threshold: int32(nb.Threshold),
		Tab: nb.Tab, CredentialName: nb.CredentialName, BastionName: nb.BastionName,
		Stats: &pb.NodeStats{
			Sites: int32(nb.Stats.Sites), Routes: int32(nb.Stats.Routes), Upstreams: int32(nb.Stats.Upstreams),
			CertCount: int32(nb.Stats.CertCount), DriftCount: int32(nb.Stats.DriftCount),
		},
		Snapshot: &pb.Snapshot{
			Id: nb.Snapshot.ID, CapturedAt: nb.Snapshot.CapturedAt, Degraded: nb.Snapshot.Degraded,
			DegradedReason: nb.Snapshot.DegradedReason, FileCount: int32(nb.Snapshot.FileCount),
			BytesRaw: nb.Snapshot.BytesRaw, ParseState: nb.Snapshot.ParseState,
		},
		HasDrift: nb.HasDrift, NoDriftReason: nb.NoDriftReason,
		SelectedSiteName: nb.SelectedSiteName, FileBody: nb.FileBody,
		ByteStart: int32(nb.ByteStart), LineStart: int32(nb.LineStart),
	}
	for _, c := range nb.Credentials {
		resp.Credentials = append(resp.Credentials, &pb.Credential{
			Id: c.ID, Name: c.Name, Username: c.Username, AuthKind: c.AuthKind,
			PublicKey: c.PublicKey, Fingerprint: c.Fingerprint, ExternalRef: c.ExternalRef, CreatedAt: c.CreatedAt,
		})
	}
	for _, in := range nb.Instances {
		resp.Instances = append(resp.Instances, toPBInstance(in))
	}
	if nb.Selected != nil {
		resp.Selected = toPBInstance(*nb.Selected)
	}
	for _, hk := range nb.HostKeys {
		resp.HostKeys = append(resp.HostKeys, &pb.HostKey{
			Id: hk.ID, NodeId: hk.NodeID, Algorithm: hk.Algorithm, Fingerprint: hk.Fingerprint,
			State: hk.State, FirstSeenAt: hk.FirstSeenAt, DecidedAt: hk.DecidedAt.String,
		})
	}
	for _, pk := range nb.PendingKeys {
		resp.PendingKeys = append(resp.PendingKeys, &pb.PendingHostKey{
			Key: &pb.HostKey{
				Id: pk.ID, NodeId: pk.NodeID, Algorithm: pk.Algorithm, Fingerprint: pk.Fingerprint,
				State: pk.State, FirstSeenAt: pk.FirstSeenAt, DecidedAt: pk.DecidedAt.String,
			},
			NodeName: pk.NodeName, NodeAddress: pk.NodeAddress, Previous: pk.Previous,
		})
	}
	for _, t := range nb.Tabs {
		resp.Tabs = append(resp.Tabs, &pb.NodeTabItem{Label: t.Label, Href: t.Href, Count: int32(t.Count), On: t.On})
	}
	resp.Inst = toPBInst(nb.Inst)
	for _, up := range nb.Upstreams {
		resp.Upstreams = append(resp.Upstreams, &pb.UpstreamPool{
			Id: up.ID, Name: up.Name, Kind: up.Kind, BalanceMethod: up.BalanceMethod,
			MemberCount: int32(up.MemberCount), UsedBy: up.UsedBy, Resolution: up.Resolution, State: up.State,
		})
	}
	if nb.SelectedPool != nil {
		resp.SelectedPool = &pb.UpstreamPool{
			Id: nb.SelectedPool.ID, Name: nb.SelectedPool.Name, Kind: nb.SelectedPool.Kind,
			BalanceMethod: nb.SelectedPool.BalanceMethod, MemberCount: int32(nb.SelectedPool.MemberCount),
			UsedBy: nb.SelectedPool.UsedBy, Resolution: nb.SelectedPool.Resolution, State: nb.SelectedPool.State,
		}
	}
	for _, m := range nb.PoolMembers {
		resp.PoolMembers = append(resp.PoolMembers, &pb.UpstreamMember{
			Host: m.Host, Port: int32(m.Port), Weight: int32(m.Weight), Flags: m.Flags,
			Role: m.Role, NodeName: m.NodeName.String,
		})
	}
	for _, c := range nb.Certificates {
		resp.Certificates = append(resp.Certificates, &pb.NodeCertBinding{
			SubjectCn: c.SubjectCN, SiteName: c.SiteName, Port: int32(c.Port), NotAfter: c.NotAfter,
			Fingerprint: c.Fingerprint, KeyType: c.KeyType, ChainLength: int32(c.ChainLength),
		})
	}
	for _, f := range nb.Files {
		resp.Files = append(resp.Files, &pb.FileRef{
			Id: f.ID, Kind: f.Kind, Path: f.Path, BytesRaw: f.BytesRaw, Truncated: f.Truncated,
		})
	}
	if nb.SelectedFile != nil {
		resp.SelectedFile = &pb.FileRef{
			Id: nb.SelectedFile.ID, Kind: nb.SelectedFile.Kind, Path: nb.SelectedFile.Path,
			BytesRaw: nb.SelectedFile.BytesRaw, Truncated: nb.SelectedFile.Truncated,
		}
	}
	for _, run := range nb.DriftRuns {
		resp.DriftRuns = append(resp.DriftRuns, &pb.DriftRun{
			Id: run.ID, InstanceId: run.InstanceID, BaselineKind: run.BaselineKind,
			BaselineLabel: run.BaselineLabel, ComputedAt: run.ComputedAt, FindingCount: int32(run.FindingCount),
		})
	}
	for _, f := range nb.DriftFindings {
		resp.DriftFindings = append(resp.DriftFindings, &pb.DriftFinding{
			Id: f.ID, ObjectKind: f.ObjectKind, NaturalKey: f.NaturalKey, Change: f.Change,
			Field: f.Field, BaselineText: f.BaselineText, SubjectText: f.SubjectText, ActionClass: f.ActionClass,
		})
	}
	if nb.DriftByObject != nil {
		resp.DriftByObject = map[string]int32{}
		for k, v := range nb.DriftByObject {
			resp.DriftByObject[k] = int32(v)
		}
	}
	resp.SelectedRoute = toPBRoute(nb.SelectedRoute)
	if nb.RouteEffect != nil {
		resp.RouteEffect = &pb.RouteEffect{
			UpstreamName: nb.RouteEffect.UpstreamName, MemberCount: int32(nb.RouteEffect.MemberCount),
			BalanceMethod: nb.RouteEffect.BalanceMethod, Members: toPBUpstreamMemberRefs(nb.RouteEffect.Members),
		}
	}
	return resp
}

func (s *Server) loadUpstreamsTab(ctx context.Context, nb *nodeBuild, snapshotID, poolID int64) error {
	pools, err := s.DB.NodeUpstreams(ctx, snapshotID)
	if err != nil {
		return err
	}
	nb.Upstreams = pools

	if poolID == 0 && len(pools) > 0 {
		poolID = pools[0].ID
	}
	for i := range pools {
		if pools[i].ID == poolID {
			nb.SelectedPool = &pools[i]
			members, err := s.DB.UpstreamMembers(ctx, poolID)
			if err != nil {
				return err
			}
			nb.PoolMembers = members
			break
		}
	}
	return nil
}

func (s *Server) loadCertificatesTab(ctx context.Context, nb *nodeBuild, snapshotID int64) error {
	query := `
		SELECT
			c.subject_cn,
			COALESCE((SELECT GROUP_CONCAT(DISTINCT sn.name)
			          FROM site_name sn WHERE sn.site_id = b.site_id), ''),
			COALESCE((SELECT l.port FROM listener l WHERE l.id = b.listener_id), 0),
			c.not_after,
			c.fingerprint_sha256,
			c.key_algorithm,
			c.key_bits,
			b.chain_depth
		FROM certificate_binding b
		JOIN certificate c ON c.id = b.certificate_id
		WHERE b.snapshot_id = ?
		ORDER BY c.not_after`

	rows, err := s.DB.R.QueryContext(ctx, query, snapshotID)
	if err != nil {
		return err
	}
	defer rows.Close()

	for rows.Next() {
		var cb nodeCertBindingBuild
		var keyAlg string
		var keyBits sql.NullInt64
		if err := rows.Scan(&cb.SubjectCN, &cb.SiteName, &cb.Port, &cb.NotAfter, &cb.Fingerprint,
			&keyAlg, &keyBits, &cb.ChainLength); err != nil {
			return err
		}
		if keyBits.Valid {
			cb.KeyType = fmt.Sprintf("%s%d", keyAlg, keyBits.Int64)
		} else if keyAlg != "" {
			cb.KeyType = keyAlg
		}
		nb.Certificates = append(nb.Certificates, cb)
	}
	return rows.Err()
}

func (s *Server) loadFilesTab(ctx context.Context, nb *nodeBuild, snapshotID, fileID int64, byteStartParam string) error {
	files, err := s.DB.SnapshotFiles(ctx, snapshotID)
	if err != nil {
		return err
	}
	nb.Files = files

	if fileID == 0 {
		return nil
	}
	for i := range nb.Files {
		if nb.Files[i].ID == fileID {
			nb.SelectedFile = &nb.Files[i]
			break
		}
	}
	if nb.SelectedFile == nil {
		return nil
	}

	body, err := s.DB.Blob(ctx, nb.SelectedFile.Digest)
	if err != nil {
		return err
	}
	content := string(body)
	nb.FileBody = content

	if byteStart, err := strconv.Atoi(byteStartParam); err == nil {
		nb.ByteStart = byteStart
		nb.LineStart = 1
		for i := 0; i < byteStart && i < len(content); i++ {
			if content[i] == '\n' {
				nb.LineStart++
			}
		}
	}
	return nil
}

func (s *Server) loadDriftTab(ctx context.Context, nb *nodeBuild) error {
	if nb.Selected == nil {
		return nil
	}
	runs, err := s.DB.DriftRuns(ctx, 0, "")
	if err != nil {
		return err
	}
	for _, run := range runs {
		if run.InstanceID == nb.Selected.ID {
			nb.DriftRuns = append(nb.DriftRuns, run)
		}
	}
	if len(nb.DriftRuns) == 0 {
		if nb.Selected.ClusterID.Valid {
			nb.NoDriftReason = "Drift detection requires a previous snapshot or cluster baseline to compare against."
		} else {
			nb.NoDriftReason = "This node is not in a cluster. Drift detection requires cluster membership or a previous snapshot."
		}
		return nil
	}

	var runIDs []int64
	for _, run := range nb.DriftRuns {
		runIDs = append(runIDs, run.ID)
		if run.FindingCount > 0 {
			nb.HasDrift = true
		}
	}
	findings, err := s.DB.DriftFindings(ctx, runIDs)
	if err != nil {
		return err
	}
	nb.DriftFindings = findings

	nb.DriftByObject = make(map[string]int)
	for _, f := range findings {
		if f.IgnoredBy.Valid {
			continue
		}
		nb.DriftByObject[f.ObjectKind]++
	}
	return nil
}

func loadRoutesTab(nb *nodeBuild, site, routePattern string) {
	if nb.Inst == nil {
		return
	}
	nb.SelectedSiteName = site

	if nb.SelectedSiteName == "" && len(nb.Inst.Sites) > 0 {
		for _, s := range nb.Inst.Sites {
			if len(s.Routes) > 0 {
				nb.SelectedSiteName = s.PrimaryName
				break
			}
		}
	}

	for _, s := range nb.Inst.Sites {
		if s.PrimaryName != nb.SelectedSiteName {
			continue
		}
		if routePattern == "" && len(s.Routes) > 0 {
			nb.SelectedRoute = s.Routes[0]
		} else {
			for _, route := range s.Routes {
				if route.Pattern == routePattern {
					nb.SelectedRoute = route
					break
				}
			}
		}
		if nb.SelectedRoute != nil {
			effect := &routeEffectBuild{}
			if nb.SelectedRoute.UpstreamID.Valid {
				if up, ok := nb.Inst.Upstreams[nb.SelectedRoute.UpstreamID.Int64]; ok {
					effect.UpstreamName = up.Name
					effect.MemberCount = len(up.Members)
					effect.BalanceMethod = up.BalanceMethod
					if effect.BalanceMethod == "" {
						effect.BalanceMethod = "round robin"
					}
					effect.Members = up.Members
				}
			}
			nb.RouteEffect = effect
		}
		break
	}
}

File: ui/src/stories/fixtures/nodes.ts
import {
  ImportNodesResponseSchema,
  NodeDetailResponseSchema,
  NodesListResponseSchema,
  TestConnectionResponseSchema,
} from '../../api/pb/nagipath/api/v1/nodes_pb'
import { pb } from '../support/mockApi'

export const nodesListFixture = pb(NodesListResponseSchema, {
  total: 428,
  threshold: 3,
  pending: 3,
  quarantined: 1,
  neverCollected: 2,
  nodes: [
    { id: 12n, address: '10.4.12.1', sshPort: 22, displayName: 'app-iad3-01', osFamily: 'linux', enabled: true, instanceCount: 2, lastCollection: '2026-08-30T09:38:00Z', lastStatus: 'succeeded', vendor: 'nginx', version: '1.24.0', cluster: 'app-iad3', processCount: 2, listeners: '443 ssl, 80', lastCaptured: '2026-08-30T09:38:00Z' },
    { id: 41n, address: '10.4.12.17', sshPort: 22, displayName: 'app-iad3-17', osFamily: 'linux', enabled: true, instanceCount: 2, lastCollection: '2026-08-30T09:35:00Z', lastStatus: 'degraded', vendor: 'nginx', version: '1.24.0', cluster: 'app-iad3', processCount: 2, listeners: '443 ssl, 80', lastCaptured: '2026-08-30T09:35:00Z' },
    { id: 88n, address: '10.9.3.4', sshPort: 22, displayName: 'app-sfo2-04', osFamily: 'linux', enabled: true, instanceCount: 1, lastCollection: '2026-08-30T09:20:00Z', lastStatus: 'succeeded', vendor: 'haproxy', version: '2.8.5', cluster: 'app-sfo2', processCount: 1, listeners: '443 ssl', lastCaptured: '2026-08-30T09:20:00Z' },
    { id: 104n, address: '10.0.4.9', sshPort: 2202, displayName: 'lb-legacy-04', osFamily: 'linux', enabled: true, consecutiveFailures: 6, instanceCount: 1, pendingHostKeys: 1, lastCollection: '2026-08-30T09:40:00Z', lastStatus: 'failed', vendor: 'apache', version: '2.4.52', cluster: 'legacy', processCount: 1, listeners: '8443 ssl', lastCaptured: '2026-08-21T02:40:00Z' },
    { id: 210n, address: '10.7.1.9', sshPort: 22, displayName: 'edge-sfo2-02', osFamily: 'linux', enabled: false, instanceCount: 0, lastStatus: '', vendor: '', cluster: 'edge-sfo2', processCount: 0, listeners: '' },
  ],
})

export const emptyNodesListFixture = pb(NodesListResponseSchema, { query: 'nothing-matches-this', threshold: 3 })

const node = {
  id: 41n,
  address: '10.4.12.17',
  sshPort: 22,
  displayName: 'app-iad3-17',
  sshUsername: 'nagipath',
  credentialId: 2n,
  osFamily: 'linux',
  sudoAvailable: true,
  enabled: true,
  source: 'import',
  notes: 'canary node for the checkout API',
  firstSeenAt: '2026-05-02T10:00:00Z',
  lastCollection: '2026-08-30T09:35:00Z',
  lastStatus: 'degraded',
}

const instances = [
  { id: 900n, nodeId: 41n, clusterId: 2n, vendor: 'nginx', displayName: 'nginx (:443)', version: '1.24.0', mainConfigPath: '/etc/nginx/nginx.conf', nodeDisplayName: 'app-iad3-17', siteCount: 31, routeCount: 214, certCount: 6, lastCaptured: '2026-08-30T09:35:00Z', clusterName: 'app-iad3', parseState: 'parsed', state: 'drift' },
  { id: 901n, nodeId: 41n, clusterId: 2n, vendor: 'nginx', displayName: 'nginx (:80)', version: '1.24.0', mainConfigPath: '/etc/nginx/nginx.conf', nodeDisplayName: 'app-iad3-17', siteCount: 31, routeCount: 34, certCount: 0, lastCaptured: '2026-08-30T09:35:00Z', clusterName: 'app-iad3', parseState: 'parsed', state: 'conforming' },
]

// Matches nodeservice.go's GetNode exactly: base + "/" + tab + "?process=" +
// selected.id — Overview has no extra path segment, every other tab does.
const tabs = [
  { label: 'Overview', href: '/nodes/41?process=900', count: 0, on: true },
  { label: 'Sites', href: '/nodes/41/sites?process=900', count: 31, on: false },
  { label: 'Routes', href: '/nodes/41/routes?process=900', count: 214, on: false },
  { label: 'Upstreams', href: '/nodes/41/upstreams?process=900', count: 8, on: false },
  { label: 'Certificates', href: '/nodes/41/certificates?process=900', count: 6, on: false },
  { label: 'Config files', href: '/nodes/41/files?process=900', count: 14, on: false },
  { label: 'Drift', href: '/nodes/41/drift?process=900', count: 4, on: false },
]

const base = {
  node,
  instances,
  selected: instances[0],
  tabs,
  credentialName: 'iad3-collector',
  threshold: 3,
  stats: { sites: 31, routes: 214, upstreams: 8, certCount: 6, driftCount: 4 },
  credentials: [
    { id: 2n, name: 'iad3-collector', username: 'nagipath', authKind: 'ssh_key', fingerprint: 'SHA256:9mQ…v1c', createdAt: '2026-05-01T08:00:00Z' },
    { id: 3n, name: 'legacy-password', username: 'root', authKind: 'password', createdAt: '2026-06-11T08:00:00Z' },
  ],
  hostKeys: [
    { id: 55n, nodeId: 41n, algorithm: 'ssh-ed25519', fingerprint: 'SHA256:2Fq…8kZ', state: 'approved', firstSeenAt: '2026-05-02T10:00:00Z', decidedAt: '2026-05-02T10:04:00Z' },
  ],
}

// Overview (design/ screen 3c).
export const nodeDetailFixture = pb(NodeDetailResponseSchema, { ...base, tab: '' })

// Routes (5a) — the two-pane site/route list with the selected route's effect.
export const nodeRoutesFixture = pb(NodeDetailResponseSchema, {
  ...base,
  tab: 'routes',
  selectedSiteName: 'checkout.example.com',
  inst: {
    id: 900n,
    nodeName: 'app-iad3-17',
    vendor: 'nginx',
    displayName: 'nginx (:443)',
    clusterName: 'app-iad3',
    sites: [
      {
        id: 300n,
        primaryName: 'checkout.example.com',
        kind: 'server',
        names: ['checkout.example.com', 'checkout-int.example.com'],
        routes: [
          { id: 4001n, matchType: 'exact', pattern: '= /healthz', ordinal: 1, isTerminal: true, targetRaw: 'return 200' },
          { id: 4002n, matchType: 'prefix', pattern: '/api/v2/', ordinal: 2, isTerminal: true, upstreamId: 700n, targetRaw: 'proxy_pass http://checkout_api_canary' },
          { id: 4003n, matchType: 'prefix', pattern: '/static/', ordinal: 3, isTerminal: true, targetRaw: 'root /srv/checkout/static' },
          { id: 4004n, matchType: 'prefix', pattern: '/', ordinal: 4, isTerminal: true, upstreamId: 701n, targetRaw: 'proxy_pass http://checkout_app' },
        ],
      },
      { id: 301n, primaryName: 'www.example.com', kind: 'server', names: ['www.example.com'], routes: [{ id: 4100n, matchType: 'prefix', pattern: '/', ordinal: 1, isTerminal: true, upstreamId: 701n, targetRaw: 'proxy_pass http://www_app' }] },
    ],
    upstreams: [
      { id: 700n, name: 'checkout_api_canary', kind: 'http', balanceMethod: 'round_robin', members: [{ id: 1n, host: '10.4.12.90', port: 8080, scheme: 'http' }] },
      { id: 701n, name: 'checkout_app', kind: 'http', balanceMethod: 'least_conn', members: [{ id: 2n, host: '10.4.12.11', port: 8080, scheme: 'http' }, { id: 3n, host: '10.4.12.12', port: 8080, scheme: 'http', flags: 'backup' }] },
    ],
  },
  selectedRoute: { id: 4002n, matchType: 'prefix', pattern: '/api/v2/', ordinal: 2, isTerminal: true, upstreamId: 700n, targetRaw: 'proxy_pass http://checkout_api_canary' },
  routeEffect: {
    upstreamName: 'checkout_api_canary',
    memberCount: 2,
    balanceMethod: 'round_robin',
    members: [
      { id: 1n, host: '10.4.12.90', port: 8080, scheme: 'http' },
      { id: 4n, host: '10.4.12.91', port: 8080, scheme: 'http', flags: 'down' },
    ],
  },
})

// Upstreams (2o) — pool table with the open pool's members underneath.
export const nodeUpstreamsFixture = pb(NodeDetailResponseSchema, {
  ...base,
  tab: 'upstreams',
  upstreams: [
    { id: 700n, name: 'checkout_api_canary', kind: 'http', balanceMethod: 'round_robin', memberCount: 2, usedBy: 'checkout.example.com /api/v2/', resolution: 'static', state: 'verified' },
    { id: 701n, name: 'checkout_app', kind: 'http', balanceMethod: 'least_conn', memberCount: 8, usedBy: 'checkout.example.com /', resolution: 'static', state: 'verified' },
    { id: 702n, name: 'www_app', kind: 'http', balanceMethod: 'round_robin', memberCount: 6, usedBy: 'www.example.com /', resolution: 'dns', state: 'inferred' },
    { id: 703n, name: 'legacy_soap', kind: 'http', balanceMethod: 'ip_hash', memberCount: 0, usedBy: '—', resolution: 'unresolved', state: 'missing' },
  ],
  selectedPool: { id: 700n, name: 'checkout_api_canary', kind: 'http', balanceMethod: 'round_robin', memberCount: 2, usedBy: 'checkout.example.com /api/v2/', resolution: 'static', state: 'verified' },
  poolMembers: [
    { host: '10.4.12.90', port: 8080, weight: 1, role: 'primary', nodeName: 'app-iad3-17' },
    { host: '10.4.12.91', port: 8080, weight: 1, flags: 'down', role: 'primary', nodeName: 'app-iad3-17' },
  ],
})

// Certificates (2p).
export const nodeCertificatesFixture = pb(NodeDetailResponseSchema, {
  ...base,
  tab: 'certificates',
  certificates: [
    { subjectCn: 'checkout.example.com', siteName: 'checkout.example.com', port: 443, notAfter: '2026-09-08T00:00:00Z', fingerprint: 'SHA256:a1b…9f0', keyType: 'ECDSA P-256', chainLength: 2 },
    { subjectCn: '*.example.com', siteName: 'www.example.com', port: 443, notAfter: '2027-02-01T00:00:00Z', fingerprint: 'SHA256:c4d…12a', keyType: 'RSA 2048', chainLength: 3 },
  ],
})

// Config files (5b) — file list plus the selected file's body.
export const nodeFilesFixture = pb(NodeDetailResponseSchema, {
  ...base,
  tab: 'files',
  snapshot: { id: 8801n, capturedAt: '2026-08-30T09:35:00Z', degraded: true, degradedReason: '2 includes unreadable', fileCount: 14, bytesRaw: 184320n, parseState: 'parsed' },
  files: [
    { id: 9101n, kind: 'main', path: '/etc/nginx/nginx.conf', bytesRaw: 4096n },
    { id: 9102n, kind: 'include', path: '/etc/nginx/conf.d/checkout.conf', bytesRaw: 2048n },
    { id: 9103n, kind: 'include', path: '/etc/nginx/conf.d/www.conf', bytesRaw: 1536n },
    { id: 9104n, kind: 'include', path: '/etc/nginx/snippets/tls.conf', bytesRaw: 512n, truncated: true },
  ],
  selectedFile: { id: 9102n, kind: 'include', path: '/etc/nginx/conf.d/checkout.conf', bytesRaw: 2048n },
  lineStart: 12,
  byteStart: 384,
  fileBody: [
    'server {',
    '    listen 443 ssl http2;',
    '    server_name checkout.example.com checkout-int.example.com;',
    '',
    '    ssl_certificate     /etc/ssl/checkout.pem;',
    '    ssl_certificate_key /etc/ssl/checkout.key;',
    '',
    '    location = /healthz { return 200; }',
    '',
    '    location /api/v2/ {',
    '        proxy_pass http://checkout_api_canary;',
    '        proxy_set_header X-Request-Id $request_id;',
    '    }',
    '',
    '    location /static/ { root /srv/checkout/static; }',
    '    location / { proxy_pass http://checkout_app; }',
    '}',
  ].join('\n'),
})

// Drift tab of the node (the per-node view of design/ screen 6b's findings).
export const nodeDriftFixture = pb(NodeDetailResponseSchema, {
  ...base,
  tab: 'drift',
  hasDrift: true,
  driftByObject: { route: 3, upstream: 1 },
  driftRuns: [
    { id: 660n, instanceId: 900n, baselineKind: 'golden_peer', baselineLabel: 'app-iad3-01', computedAt: '2026-08-30T09:36:00Z', findingCount: 4, instanceName: 'nginx (:443)', nodeName: 'app-iad3-17', vendor: 'nginx', parserVersion: 7, ignoredCount: 1 },
  ],
  driftFindings: [
    { id: 7701n, objectKind: 'route', naturalKey: 'checkout.example.com /api/v2/', change: 'modified', field: 'upstream', baselineText: 'checkout_api', subjectText: 'checkout_api_canary', actionClass: 'proxy', provenance: { path: '/etc/nginx/conf.d/checkout.conf', snapshotId: 8801n, fileId: 9102n, byteStart: 384n, link: '/snapshots/8801/file/9102?b=384' } },
    { id: 7702n, objectKind: 'upstream', naturalKey: 'checkout_api_canary', change: 'added', field: '', subjectText: '2 members', actionClass: 'upstream', provenance: { path: '/etc/nginx/conf.d/checkout.conf', snapshotId: 8801n, fileId: 9102n, byteStart: 1024n, link: '/snapshots/8801/file/9102?b=1024' } },
  ],
})

export const importNodesFixture = pb(ImportNodesResponseSchema, {
  added: [
    { id: 431n, displayName: 'lb01.example.com', address: 'lb01.example.com', sshPort: 22 },
    { id: 432n, displayName: 'web01', address: '10.90.4.2', sshPort: 22 },
    { id: 433n, displayName: 'db02.example.com', address: 'db02.example.com', sshPort: 22 },
    { id: 434n, displayName: 'edge-haproxy-03', address: 'edge-haproxy-03', sshPort: 22 },
  ],
  refused: [
    'line 6: "10.90.4.0/24" — a range is not a host',
    'line 9: "db01.example.com" — already in inventory',
  ],
})

// One /nodes/{id}/test fixture per node in importNodesFixture.added above, so
// step 2 of the Import inventory story shows every connection-test outcome at
// once: connected, blocked on an unapproved host key, and failed.
export const importTestConnectedFixture = pb(TestConnectionResponseSchema, {
  status: 'connected', latencyMs: 41, osFamily: 'linux',
})
export const importTestHostKeyPendingFixture = pb(TestConnectionResponseSchema, {
  status: 'host_key_pending',
  pendingKey: { id: 901n, nodeId: 432n, algorithm: 'ed25519', fingerprint: 'SHA256:pR7g…4Nc', state: 'pending', firstSeenAt: '2026-08-31T04:31:00Z' },
})
export const importTestFailedFixture = pb(TestConnectionResponseSchema, {
  status: 'failed', error: 'dial tcp 10.9.4.4:22: connect: connection refused',
})
